### ISSUE TITLE
Plugins fail on non-interactive composer create-project

### ISSUE THREAD

**Issue by davidwindell** [composer/composer#10928]
I can no longer install Magento with `composer create-project --no-interaction`. It seems to affect composer back to v2.3.0 at least (not tested 2.2 but it was working fine only a few weeks ago).

The command I run to install it is:

```
composer create-project --no-interaction --prefer-dist --no-dev --repository=https://repo.magento.com/ magento/project-community-edition /var/www 2.4.3-p2
```

**Comment by liszkapawel** [comment]
After the new composer version is released, I have the same problem with each of our Sylius plugins in github actions:

```
Error: symfony/flex contains a Composer plugin which is blocked by your allow-plugins config. You may add it to the list if you consider it safe.
You can run "composer config --no-plugins allow-plugins.symfony/flex [true|false]" to enable it (true) or disable it explicitly and suppress this exception (false)
See https://getcomposer.org/allow-plugins
In PluginManager.php line 762:
  symfony/flex contains a Composer plugin which is blocked by your allow-plug  
  ins config. You may add it to the list if you consider it safe.              
  You can run "composer config --no-plugins allow-plugins.symfony/flex [true|  
  false]" to enable it (true) or disable it explicitly and suppress this exce  
  ption (false)                                                                
  See https://getcomposer.org/allow-plugins
```

Adding allow-plugins in composer.json like that does not solve the problem:
```
"allow-plugins": {
            "symfony/flex": true,
        }
```

**Comment by stof** [comment]
@davidwindell if the magento skeleton includes some plugins by default, it should probably configure them as allowed or no in the skeleton, to simplify the experience of new users (see https://github.com/symfony/skeleton/pull/201 for the example about the Symfony skeleton).

Otherwise, it forces to create the project with `--no-install`, then configure plugins as allowed or no and then run `composer install`

**Comment by davidwindell** [comment]
I guess we've no way to address this then as I can't see us being able to influence Magento's skeleton right now.

The best workaround I could come up with for now was to create a temporary `composer.json` and use `composer require` instead like this https://github.com/outeredge/edge-docker-magento/commit/e73840ea73b135bc8a6e6b3c6c66401c3940d1a3, but it would be so much better if we could go back to the old way and have the option to say `--allow-all-plugins` or something like that when using `composer create-project`

**Comment by hostep** [comment]
The Magento skeleton file got already updated by https://github.com/magento/magento2/pull/34873
But they decided to only include these changes in Magento 2.4.4. So unfortunately not for versions 2.3.7-p3 or 2.4.3-p2 which are also still currently supported versions that were released a long time after composer 2.2.x was released.

But @stof's workaround is probably going to work:
> Otherwise, it forces to create the project with --no-install, then configure plugins as allowed or no and then run composer install

**Comment by stof** [comment]
For a plugin installed globally, you must configure it as allowed or no in the global composer config, not in the local config.

**Comment by Wotuu** [comment]
Hi stof, we tried reverting back to 2.3.7 but had troubles doing that - running the command `composer global config --no-plugins allow-plugins.hirak/prestissimo true` as stated in the error message does resolve the issue. I'm sure changing the global config would've also worked but we stopped troubleshooting as we found this workaround. Will circle back to this later to see what we can structurally do about it. Thanks again.

**Comment by davidwindell** [comment]
Yeah, I found the same, no matter what global config allows I set it still wouldn't install non-interactively.

**Comment by davidwindell** [comment]
This is caused by the July 2022 enforced plugin changes (regardless of
version).

**Comment by kayacekovic** [comment]
you can add `composer global config --no-plugins allow-plugins.hirak/prestissimo false` to top of ci/cd or deploy file. its worked for me.

via https://github.com/ahmetkorkmaz3

![image](https://user-images.githubusercontent.com/32109756/177949306-311c8cf4-140d-43a4-96e3-6a556e5f9f07.png)

**Comment by MPa1977** [comment]
> I can no longer install Magento with `composer create-project --no-interaction`. I see this is caused by the July 2022 change.
> 
> The command I run to install it is:
> 
> ```
> composer create-project --no-interaction --prefer-dist --no-dev --repository=https://repo.magento.com/ magento/project-community-edition /var/www 2.4.3-p2
> ```
> 
> Leads to:
> 
> ```
> #14 59.75 In PluginManager.php line 762:
> #14 59.75                                                                                
> #14 59.75   laminas/laminas-dependency-plugin contains a Composer plugin which is block  
> #14 59.75   ed by your allow-plugins config. You may add it to the list if you consider  
> #14 59.75    it safe.                                                                    
> #14 59.75   You can run "composer config --no-plugins allow-plugins.laminas/laminas-dep  
> #14 59.75   endency-plugin [true|false]" to enable it (true) or disable it explicitly a  
> #14 59.75   nd suppress this exception (false)                                           
> #14 59.75   See https://getcomposer.org/allow-plugins 
> ```
> 
> How can I use this command in a Docker build?
This is what worked for my docker build

`RUN set -eux; \
      composer global config --no-plugins allow-plugins.symfony/flex true; \
      composer global require "symfony/flex" --prefer-dist --no-progress --no-suggest --classmap-authoritative; \
	  composer clear-cache`

**Comment by cmuench** [comment]
We have the same issue in the n98-magerun2 build pipeline. Seems that our install command is now broken. In the background we run `composer create-project`.

**Comment by MaxChri** [comment]
This composer shit fucks up. Why do they have to change something nobody asked for.

**Comment by andytson-inviqa** [comment]
the documentation on https://getcomposer.org/doc/06-config.md#allow-plugins conflicts with actual behaviour. The documentation says it will warn on seeing a new plugin, instead it fatally errors
> When a new plugin is first activated, which is not yet listed in the config option, Composer will print a warning.

**Comment by ihor-sviziev** [comment]
We're having quite a similar issue with command 
```
composer create-project magento/magento-coding-standard magento-coding-standard 16.*
```
on Composer 2.3.7 - worked fine ✅
<img width="1053" alt="image" src="https://user-images.githubusercontent.com/1873745/178316880-22f13431-4282-4106-b77f-5c8f33c8f792.png">


After updating to 2.3.9 - it started to fail: ❌
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/1873745/178316678-ae5cb5b3-2ce5-4457-b365-2db1f226d89d.png">


Looks like adding `--no-plugins` flag does the trick, but I can expect such breaking changes in minor version, not in the patch release.

Looks like the changes were caused by https://github.com/composer/composer/pull/10920. I would prefer to have them in Composer 2.4, not in 2.3.9.

**Comment by andytson-inviqa** [comment]
3.0, as a BC break

**Comment by ihor-sviziev** [comment]
Hi @Seldaek,
What do you think about reverting backward incompatible change in 2.3.x release line and add it to 2.4.0 or 3.0? Does it make sense?

**Comment by hostep** [comment]
For the people having trouble with setting up Magento using `composer create-project` (with versions lower than Magento 2.4.4), the following seems to work without problems with Composer version 2.3.9:
```bash
$ composer create-project --no-install --repository-url=https://repo.magento.com/ magento/project-community-edition=2.4.3-p2 .
$ composer config allow-plugins.dealerdirect/phpcodesniffer-composer-installer true
$ composer config allow-plugins.laminas/laminas-dependency-plugin true
$ composer config allow-plugins.magento/* true
$ composer install
```

The trick lies in the `--no-install` flag, like @stof [mentioned before](https://github.com/composer/composer/issues/10928#issuecomment-1177265009).

**Comment by cmuench** [comment]
> For the people having trouble with setting up Magento using `composer create-project` (with versions lower than Magento 2.4.4), the following seems to work without problems with Composer version 2.3.9:
> 
> ```shell
> $ composer create-project --no-install --repository-url=https://repo.magento.com/ magento/project-community-edition=2.4.3-p2 .
> $ composer config allow-plugins.dealerdirect/phpcodesniffer-composer-installer true
> $ composer config allow-plugins.laminas/laminas-dependency-plugin true
> $ composer config allow-plugins.magento/* true
> $ composer install
> ```
> 
> The trick lies in the `--no-install` flag, like @stof [mentioned before](https://github.com/composer/composer/issues/10928#issuecomment-1177265009).
> 
> I don't have anything against this new behavior of composer, since it was announced at least half a year before it changed.

I was aware of this but this requires some rewrites in n98-magerun2 install command. Nevertheless, it seems that this is now the new default behavior which makes the world a bit more complex if we need to install a older version.

**Comment by davidwindell** [comment]
Surely `create-project` implies you are happy to accept any plugins from the project you are creating from?

**Comment by stof** [comment]
Also, they could include their list of allowed plugins in their project skeleton to fix that for their user for the plugins intended to be included by default. That's what other projects have done. And Magento even did that for their upcoming version. But they haven't applied it to the skeletons for their stable versions.

**Comment by Seldaek** [comment]
For people having issues with hirak/prestissimo, I would recommend you remove it (`composer global remove hirak/prestissimo --no-plugins` probably does the trick) as it is completely useless and does not do anything if you are running Composer 2.x.

That said, it is a regression that it fails on this plugin as it should simply be skipped on Composer 2. I have fixed that now in https://github.com/composer/composer/commit/8323e85bc844466fc71a85782bbd0f6141aa2750

**Comment by Seldaek** [comment]
Some of the create-project issues which happened only when running composer in an existing project's dir are resolved by https://github.com/composer/composer/commit/75ef4903ef70d1e95028d0391b1f5ca8f3a4e1e1 - please try with `composer self-update --snapshot` see if that improves things.

The issue where the project created requires plugins itself and does not declare allow-plugins however is not fixed as per my comment above.

I'll try to get a release out tomorrow with this and a few more fixes.

**Comment by Seldaek** [comment]
Added a hint in https://github.com/composer/composer/commit/336a0d20c6cd2004b2374f88c30cd15bece7d0ba to suggest the workaround, and added the workaround to the OP here as well to help anyone googling to hopefully find the solution quickly.

Closing this issue now as I don't see what else we can do. If someone still has unexpected behavior please report a new issue.

BTW [2.3.10](https://github.com/composer/composer/releases/tag/2.3.10) is now released with all the above fixes.
